docs/bugpattern/UnnecessaryLambda.md

```diff
@@ -34,3 +34,17 @@ Advantages of using a method include:
 *   If the behavior is nontrivial, it's more natural to write javadoc for the
     method.
 
+Be aware that this change is not purely syntactic: it affects the semantics of
+your program in some small ways. In particular, evaluating the same method
+reference twice is not guaranteed to return an identical object.
+
+This means that, first, inlining the reference instead of storing a lambda may
+cause additional memory allocations - usually this very slight performance cost
+is worth the improved readability, but use your judgment if the performance
+matters to you.
```

Comment:
@amalloy I'm lost by these docs for a couple of reasons.

Firstly, I can't see why changing a stored lambda to a method reference would in of itself cause a lack of a guarantee to return an identical object, because AFAIU stored lambdas have the exact same problem...

...And secondly, I'm struggling to tell why using a method reference over a stored lambda would cause more memory to be used. 🤔

Would you mind expanding on these in the docs?

Reply:
Consider

```java
private IntSupplier hash = this::hashCode;
public boolean same() {
  return hash == hash;
}
```

The comparison is guaranteed to be true, and we will allocate exactly one lambda object over the lifetime of this object.

Now apply the rewrite rule, yielding:

```java
public boolean same() {
  return this::hashCode == this::hashCode;
}
```

which is not guaranteed to be the same, because two different lambdas may be allocated for each call to `same`. This is the difference the new paragraphs warn about. Perhaps it would be worth adding this example to the documentation.

Reply:
Ah okay, so am I right to understand that in the second code sample, the two `this::hashCode` lambdas aren't guaranteed to be equal to one another because the JVM could assign them different object identities?

This wasn't clear to me from the docs, because I thought they were talking about the lambdas' _outputs_ rather than the lambdas themselves. So, although I don't know how one could better word the docs, I agree that it would be worth including this example to help prevent confusion. :)

Reply:
This would also explain to me why the refactoring would increase memory usage slightly, because naturally if both lambdas have different object identities, then we end up not with _one_ object but _two_ of them.